token/program-2022/src/extension/reallocate.rs

```diff
@@ -72,19 +74,41 @@ pub fn process_reallocate(
     // if additional lamports needed to remain rent-exempt, transfer them
     let rent = Rent::get()?;
     let new_minimum_balance = rent.minimum_balance(needed_account_len);
-    let lamports_diff = new_minimum_balance.saturating_sub(token_account_info.lamports());
-    invoke(
-        &system_instruction::transfer(payer_info.key, token_account_info.key, lamports_diff),
-        &[
-            payer_info.clone(),
-            token_account_info.clone(),
-            system_program_info.clone(),
-        ],
-    )?;
+
+    let current_lamport_reserve = token_account_info
+        .lamports()
+        .checked_sub(native_amount.unwrap_or(0))
+        .ok_or(TokenError::Overflow)?;
+    let lamports_diff = new_minimum_balance.saturating_sub(current_lamport_reserve);
+    if lamports_diff > 0 {
+        invoke(
+            &system_instruction::transfer(payer_info.key, token_account_info.key, lamports_diff),
+            &[
+                payer_info.clone(),
+                token_account_info.clone(),
+                system_program_info.clone(),
+            ],
+        )?;
+    }
 
     // unpack to set account_type, if needed
-    let mut token_account = token_account_info.data.borrow_mut();
-    set_account_type::<Account>(&mut token_account)?;
+    let mut token_account_data = token_account_info.data.borrow_mut();
+    set_account_type::<Account>(&mut token_account_data)?;
+
+    // sync the rent exempt reserve for native accounts
+    let mut token_account = StateWithExtensionsMut::<Account>::unpack(&mut token_account_data)?;
+    if let Some(native_amount) = native_amount {
+        // sanity check that there are enough lamports to cover the token amount
+        // and the rent exempt reserve
+        let lamports_in_account_data = native_amount
+            .checked_add(new_minimum_balance)
```

Comment:
nit: `new_minimum_balance.checked_add(native_amount)` for consistency with L82's `.saturating_sub(...)`.

Reply:
The `saturating_sub` makes sense because you could have too many lamports in your account, but I did a `checked_add` here in the weird case where someone overallocates a huge account and also has a lot of lamports, but you're right, it's fine (and simpler) to saturate.

Reply:
Ah ok, that makes way more sense! I went back to checked and changed the order